src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/ConversionsBase.cs

```diff
@@ -2504,25 +2528,27 @@ public bool HasBoxingConversion(TypeSymbol source, TypeSymbol destination, ref H
 
             if (source.IsRestrictedType())
             {
-                return false;
+                return ConversionKind.NoConversion;
             }
 
             if (destination.Kind == SymbolKind.DynamicType)
             {
-                return !source.IsPointerType();
+                return !source.IsPointerType()
+                    ? ConversionKind.ValueTypeBoxing
+                    : ConversionKind.NoConversion;
             }
 
             if (IsBaseClass(source, destination, ref useSiteDiagnostics))
             {
-                return true;
+                return ConversionKind.ValueTypeBoxing;
             }
 
             if (HasAnyBaseInterfaceConversion(source, destination, ref useSiteDiagnostics))
             {
-                return true;
+                return ConversionKind.ValueTypeBoxing;
```

Comment:
This looks like this could be wrong.  After all, an type parameter coudl have a base interface that it was being boxed to.  However, that case is handled in:

```c#
if ((source.TypeKind == TypeKind.TypeParameter) &&
                  HasImplicitBoxingTypeParameterConversion((TypeParameterSymbol)source, destination, ref useSiteDiagnostics))
```

Note: I could add an assert here that this shouldn't return true if source was a type parameter because of that (at least if my logic is correct).